File: Examples/ControlPointDemo_ObjC/ControlPointDemo/FolderViewController.m
//
//  FolderViewController.m
//
//  Copyright (c) 2015 David Robles
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.

#import "FolderViewController.h"
#import "Player.h"
@import UPnAtom;

@interface FolderViewController () <UITableViewDataSource, UITableViewDelegate>
@property (nonatomic) IBOutlet UITableView *tableView;
@end

@implementation FolderViewController {
    NSString *_rootId;
    MediaServer1Device *_device;
    NSArray *_playlist;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [_device.contentDirectoryService getSortCapabilities:^(NSString *sortCaps) {
        if ([sortCaps rangeOfString:@"dc:title"].location != NSNotFound) {
            sortCaps = @"+dc:title";
        }
        
        [_device.contentDirectoryService browseWithObjectID:_rootId browseFlag:@"BrowseDirectChildren" filter:@"*" startingIndex:@"0" requestedCount:@"0" sortCriteria:sortCaps success:^(NSArray *result, NSInteger numberReturned, NSInteger totalMatches, NSString *updateID) {
            _playlist = result;
            [self.tableView reloadData];
        } failure:^(NSError *error) {
            NSLog(@"failed to browse content directory");
        }];
    } failure:^(NSError *error) {
        NSLog(@"failed to get sort capabilities");
    }];
    
    CGFloat viewWidth = self.navigationController.view.frame.size.width;
    UILabel *titleLabel = [[UILabel alloc] initWithFrame:CGRectMake(0.0 , 11.0f, viewWidth - (viewWidth * 0.2), 21.0f)];
    [titleLabel setFont:[UIFont fontWithName:@"Helvetica" size:18]];
    [titleLabel setBackgroundColor:[UIColor clearColor]];
    [titleLabel setTextColor:[UIColor blackColor]];
    
    if([[Player sharedInstance] mediaRenderer] == nil){
        [titleLabel setText:@"No Renderer Selected"];
    }else{
        [titleLabel setText:[[[Player sharedInstance] mediaRenderer] friendlyName] ];
    }
    
    [titleLabel setTextAlignment:NSTextAlignmentLeft];
    UIBarButtonItem *title = [[UIBarButtonItem alloc] initWithCustomView:titleLabel];
    NSArray *items = @[[[Player sharedInstance] playPauseButton], [[Player sharedInstance] stopButton], title];
    self.toolbarItems = items;
    
    self.navigationController.toolbarHidden = NO;
}

- (void)configureWithDevice:(MediaServer1Device *)device header:(NSString *)header rootId:(NSString *)rootId{
    _device = device;
    _rootId = rootId;
    self.title = header;
    _playlist = [NSArray array];
}

#pragma mark - UITableViewDataSource methods

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [_playlist count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *cellIdentifier = @"DefaultCell";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    
    // Configure the cell...
    ContentDirectory1Object *item = _playlist[indexPath.row];
    [[cell textLabel] setText:[item title]];
    
    cell.accessoryType = item.isContentDirectory1Container ? UITableViewCellAccessoryDisclosureIndicator : UITableViewCellAccessoryNone;
    
    return cell;
}

#pragma mark - UITableViewDelegate methods

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    ContentDirectory1Object *item = _playlist[indexPath.row];
    if([item isContentDirectory1Container]){
        ContentDirectory1Container *container = _playlist[indexPath.row];
        UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
        FolderViewController *targetViewController = [storyboard instantiateViewControllerWithIdentifier:@"FolderViewControllerScene"];
        [targetViewController configureWithDevice:_device header:[container title] rootId:[container objectID]];
        
        [[self navigationController] pushViewController:targetViewController animated:YES];
    }else{
        [[Player sharedInstance] startPlayback:_playlist position:indexPath.row];
    }
}

@end

File: Examples/ControlPointDemo_Swift/ControlPointDemo/FolderViewController.swift
//
//  FolderViewController.swift
//  ControlPointDemo
//
//  Created by David Robles on 3/14/15.
//  Copyright (c) 2015 David Robles. All rights reserved.
//

import UIKit
import UPnAtom

class FolderViewController: UIViewController {
    @IBOutlet fileprivate weak var _tableView: UITableView!
    fileprivate var _playlist = [ContentDirectory1Object]()
    fileprivate var _mediaServer: MediaServer1Device!
    fileprivate var _contentDirectoryID: String!
    
    func configure(mediaServer: MediaServer1Device, title: String, contentDirectoryID: String) {
        _mediaServer = mediaServer
        _contentDirectoryID = contentDirectoryID
        self.title = title
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        _mediaServer?.contentDirectoryService?.getSortCapabilities({ (sortCapabilities: String?) -> Void in
            var sortCapabilities = sortCapabilities != nil ? sortCapabilities! : ""
            if sortCapabilities.range(of:"dc:title") != nil {
                sortCapabilities = "+dc:title"
            }
            
            self._mediaServer?.contentDirectoryService?.browse(objectID: self._contentDirectoryID, browseFlag: "BrowseDirectChildren", filter: "*", startingIndex: "0", requestedCount: "0", sortCriteria: sortCapabilities, success: { (result: [ContentDirectory1Object], numberReturned: Int, totalMatches: Int, updateID: String?) -> Void in
                self._playlist = result
                self._tableView.reloadData()
                }, failure: { (error: NSError) -> Void in
                    print("Failed to browse content directory: \(error)")
            })
            }, failure: { (error: NSError) -> Void in
            print("Failed to get sort capabilities: \(error)")
        })
        
        let viewWidth = self.navigationController!.view.frame.size.width
        let titleLabel = UILabel(frame: CGRect(x: 0.0, y: 11.0, width: viewWidth - (viewWidth * 0.2), height: 21.0))
        titleLabel.font = UIFont(name: "Helvetica", size: 18)
        titleLabel.backgroundColor = UIColor.clear
        titleLabel.textColor = UIColor.black
        titleLabel.textAlignment = .left
        titleLabel.text = Player.sharedInstance.mediaRenderer == nil ? "No Renderer Selected" : Player.sharedInstance.mediaRenderer?.friendlyName
        let barButton = UIBarButtonItem(customView: titleLabel)
        self.toolbarItems = [Player.sharedInstance.playPauseButton, Player.sharedInstance.stopButton, barButton]
        
        self.navigationController?.isToolbarHidden = false
    }
}

extension FolderViewController: UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return _playlist.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell: UITableViewCell = tableView.dequeueReusableCell(withIdentifier: "DefaultCell") as UITableViewCell!
        let item = _playlist[indexPath.row]
        cell.textLabel?.text = item.title
        cell.accessoryType = item is ContentDirectory1Container ? .disclosureIndicator : .none
        
        return cell
    }
}

extension FolderViewController: UITableViewDelegate {
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let item = _playlist[indexPath.row]
        if let containerItem = item as? ContentDirectory1Container {
            let targetViewController = UIStoryboard(name: "Main", bundle: nil).instantiateViewController(withIdentifier: "FolderViewControllerScene") as! FolderViewController
            targetViewController.configure(mediaServer: _mediaServer, title: containerItem.title, contentDirectoryID: containerItem.objectID)
            self.navigationController?.pushViewController(targetViewController, animated: true)
        }
        else {
            Player.sharedInstance.startPlayback(_playlist, position: indexPath.row)
        }
    }
}
